packages/cpp/src/node/debug/cpp-debug-contribution.ts

```diff
@@ -0,0 +1,153 @@
+/********************************************************************************
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+const path = require('path');
+const packageJson = require('../../../package.json');
+const debugAdapterDir = packageJson['debugAdapter']['dir'];
+
+import { injectable } from 'inversify';
+import { DebugConfiguration } from '@theia/debug/lib/common/debug-common';
+import { DebugAdapterContribution, DebugAdapterExecutable } from '@theia/debug/lib/node/debug-model';
+import * as Ajv from 'ajv';
+
+export const CppDebugConfigurationSchema = require('../../../data/cpp-debug-configuration-schema.json');
+export namespace cppDebugConfigurationValidators {
+    export const Launch = new Ajv().compile(CppDebugConfigurationSchema['launch']);
+    export const Attach = new Ajv().compile(CppDebugConfigurationSchema['attach']);
+}
+
+@injectable()
+export class GdbDebugAdapterContribution implements DebugAdapterContribution {
+
+    readonly debugType = 'gdb';
+    readonly filePatterns = [
+        '[.]c$',
+        '[.]cpp$',
+        '[.]d$',
+        '[.]objective-c$',
+        '[.]fortran$',
+        '[.]fortran-modern$',
+        '[.]fortran90$',
+        '[.]fortran_free-form$',
+        '[.]fortran_fixed-form$',
+        '[.]rust$',
+        '[.]pascal$',
+        '[.]objectpascal$',
+        '[.]ada$',
+        '[.]nim$',
+        '[.]arm$',
+        '[.]asm$',
+        '[.]vala$',
+        '[.]crystal$',
+    ];
+
+    provideDebugConfigurations = [
+        {
+            type: 'gdb',
+            request: 'launch',
+            name: 'Launch Program',
+            target: './bin/executable',
+            cwd: '${workspaceRoot}',
```

Comment:
Does expansion of `${workspaceRoot}` work?

Reply:
I currently don't know a way to resolve variables from the backend, so no.

Although it should be a thing.

Should I remove the template-like placeholder for now?

Reply:
I think so. Perhaps create an issue to be able to do that? It would be very useful.

Reply:
Actually after skimming through the code, there seems to be support for config resolution, but let me check if this actually works...

Reply:
Turns out the Theia equivalent for vscode's `workspaceRoot` is `workspaceFolder`. But it works :)

Reply:
Just found this too, that's awesome!